Core/src/com/serotonin/m2m2/rt/script/CompiledMangoJavaScript.java

```diff
@@ -0,0 +1,232 @@
+import javax.script.ScriptEngine;
+import javax.script.ScriptException;
+import javax.script.SimpleBindings;
+
+import com.infiniteautomation.mango.spring.service.MangoJavaScriptService;
+import com.infiniteautomation.mango.util.script.MangoJavaScriptResult;
+import com.infiniteautomation.mango.util.script.ScriptPermissions;
+import com.infiniteautomation.mango.util.script.ScriptUtility;
+import com.serotonin.ShouldNeverHappenException;
+import com.serotonin.db.pair.IntStringPair;
+import com.serotonin.m2m2.Common;
+import com.serotonin.m2m2.module.MangoJavascriptContextObjectDefinition;
+import com.serotonin.m2m2.module.ModuleRegistry;
+import com.serotonin.m2m2.rt.dataImage.IDataPointValueSource;
+import com.serotonin.m2m2.rt.dataImage.PointValueTime;
+import com.serotonin.m2m2.rt.dataImage.types.DataValue;
+import com.serotonin.m2m2.vo.permission.PermissionHolder;
+
+/**
+ * Container for Compiled Java Scripts to aid in state management 
+ * of the various scopes and to help simplify user code.
+ * 
+ * Usage:
+ * 1. Constructe compiled script
+ * 2. call compile()
+ * 3. call initialize() (anytime the context changes)
+ * 4. call execute() when you desire to compute a result from the script
+ * 
+ * @author Terry Packer
+ *
+ */
+public class CompiledMangoJavaScript {
+    
+    private CompiledScript compiledScript;
+    
+    private final List<ScriptUtility> utilities;
+    private Map<String, IDataPointValueSource> context;
+    
+    private final Map<String, Object> additionalContext;
+    private final List<ScriptUtility> additionalUtilities;
+    private ScriptPermissions permissions;
+    private ScriptLog log;
+    private ScriptPointValueSetter setter;
+    private List<JsonImportExclusion> importExclusions;
+    
+    
+    private final boolean testRun;
+    private final MangoJavaScriptService service;
+    private final PermissionHolder user;
+    private final MangoJavaScriptResult result;
+    
+    
+    public CompiledMangoJavaScript(Map<String, Object> additionalContext, List<ScriptUtility> additionalUtilities, boolean testRun, MangoJavaScriptService service, PermissionHolder user) {
+        this.additionalContext = additionalContext == null ? Collections.emptyMap() : additionalContext;
+        this.additionalUtilities = additionalUtilities == null ? Collections.emptyList() : additionalUtilities;
+        this.testRun = testRun;
+        this.service = service;
+        this.user = user;
+        this.result = new MangoJavaScriptResult();
+        this.utilities = new ArrayList<>();
+    }
+    
+    
+    public void compile(String script, boolean wrapInFunction) throws ScriptError {
+        try {
+            final ScriptEngine engine = service.newEngine(user);
+            
+            //TODO Should these REALLY be globally scoped?
+            Bindings globalBindings = new SimpleBindings();
+            // Add constants to the context.
+            globalBindings.put("SECOND", Common.TimePeriods.SECONDS);
```

Comment:
Any reason not to use TIME_PERIOD_CODES like ROLLUP_CODES is used below, just excluding milliseconds (or not?)

Reply:
Seems that the ExportCodes have an 's' on the end.  I.e. SECOND is added here but the export code is SECONDS.  

I see the merit here so I'm going to leave these here and also add the Export code values.